### ISSUE TITLE
Functoria: need a way to disable Bootvar

### ISSUE THREAD

**Issue by talex5** [mirage/mirage#493]
Is there a way to disable the `Bootvar` parsing? Some platforms don't give you control of what kernel arguments you get, and if functoria can't parse them then the unikernel is unbootable. Qubes, for example, passes lots of arguments intended for Linux. As a work-around, I manually edit my `main.ml` after running `mirage` to:

    let argv_xen1 = lazy (return (`Ok [| "" |]))

**Comment by Drup** [comment]
Hum, no, there isn't. I'm curious though, wouldn't argv be empty in this case ? If it's not the case, shouldn't bootvar-xen be fixed ? It doesn't seem very good to already disable bootvars during configuration.

**Comment by Drup** [comment]
If you insist that being able to disable bootvar at configure time, it should be very easy to add one more argv device, that always returns `""` and hide the whole thing behind an `if_impl`.

**Comment by talex5** [comment]
It doesn't like it being empty for some reason:

    MirageOS booting...
    Initialising timer interface
    Initialising console ... done.
    Netif: add resume hook
    gnttab_stubs.c: initialised mini-os gntmap
    qubes-firewall: internal error, uncaught exception:
		    Invalid_argument("argv array must have at least one element")
		    Raised at file "pervasives.ml", line 31, characters 25-45
    Mirage exiting with status 1
    Do_exit called!

**Comment by Drup** [comment]
Oh, that is indeed a bug...

- Cmdliner removes the first element of argv (and returns this error when argv is empty).
- mirage-bootvar-xen doesn't always add the executable name at the beginning.

Which also means it's probably going to ignore the first options ....

@MagnusS Your opinion ? Should I add the executable name myself in functoria, or should it be done in bootvar ?

**Comment by MagnusS** [comment]
Since we now have `argv` and `to_argv` in bootvar I think we should fix them to work as expected and always set first entry to kernel name or a placeholder.

**Comment by Drup** [comment]
Actually, I just looked in more details, and we do add one empty string at the beginning of the array
@talex5 How did you triggered that ?

**Comment by talex5** [comment]
If you configure with `--unix` it generates:

    OS.Env.argv () >>= (fun x -> Lwt.return (`Ok x))

`OS.Env.argv ()` is just `Lwt.return (Sys.argv)`, so it looks like it does expect the program name here.

You should be able to reproduce it with any unikernel by editing the `main.ml` to give an empty array and then running it.

**Comment by Drup** [comment]
> You should be able to reproduce it with any unikernel by editing the main.ml to give an empty array and then running it.

Right, so you are just replacing the generated code by faulty code. There is nothing to fix here, just don't give an empty array ....

More important, what is the failure in the original case ("Some platforms don't give you control of what kernel arguments you get, and if functoria can't parse them then the unikernel is unbootable.")

**Comment by talex5** [comment]
Yes, I know. I was just responding to your question about why I used `[| "" |]` instead of `[| |]`.

The issue is that some systems (e.g. Qubes) supply a whole load of extra arguments (intended for Linux systems), which functoria can't parse. This makes the unikernels unbootable on such systems. It's not really functoria's fault, but we probably should provide a work-around.

**Comment by Drup** [comment]
Right, we just misunderstood each other. I understand the issue better now. I see two solutions.

- Create a device that accepts qubes parameters
- Create an empty_arvg device, as proposed earlier.

**Comment by talex5** [comment]
Yes, sorry - I see now you meant: why doesn't Qubes pass an empty list? It actually gives us:

    root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH 3

Is there a way to specify an `empty_argv` device in `config.ml` now, or will it require changes to mirage?

**Comment by Drup** [comment]
It will require changes to mirage.

**Comment by Drup** [comment]
@talex5 Does bootvar even parses this command line ? It looks like you would have the same issue than https://github.com/mirage/mirage-bootvar-xen/issues/17

**Comment by talex5** [comment]
I haven't got my test machine with me at the moment but I think you're right: bootvar complains that `ro` is malformed because it doesn't contain an `=`.

**Comment by MagnusS** [comment]
The current parsing in bootvar is limited - it was originally only intended to be used for key/value pairs, not command line options. The current `argv` function makes it work with functoria, but doesn't always behave as expected. For example, space, quotes and keys without values are not allowed and parameters get '-' and '--' automatically prepended. We can improve the regexp, but to fix it properly we probably need to parse the option string.

See also discussion about improving the parser here: https://github.com/mirage/mirage-bootvar-xen/pull/14

**Comment by Drup** [comment]
Let's assume that all the parsing problems can be fixed in bootvar (I'm very confident we can do it!)

> Good point. Qubes at least allows you to add additional arguments (but not get rid of the default ones). Perhaps we could pass some marker - e.g. get Bootvar to ignore everything before --? (Comment from the other thread)

This is an interesting idea for when you have no idea where you are going to deploy. However, I think a better solution would be to create a device that accept the Qubes parameters as keys, which would give you access to those parameters inside your unikernel. Same thing for EC2.

**Comment by Drup** [comment]
And you can already define this device (and the associated keys) inside your config.ml.

**Comment by talex5** [comment]
@Drup could you give an example showing how to override the default argv handling in a `config.ml`?

**Comment by Drup** [comment]
You don't override anything. You just create new keys. For example, to be able to accept `root=/dev/mapper/dmroot`: 

```ocaml
let root_key = 
  let doc = Key.Arg.info ~doc:"Root of something." ["root"]  in
  Key.(create "root" Arg.(opt (some string) None doc))
```

Then you add it to some device, or directly to the register function:
```ocaml
  ~keys:Key.[abstract root_key]
```

**Comment by talex5** [comment]
And I have to do this for every key I don't care about, to stop functoria crashing with a `Usage` error?

**Comment by Drup** [comment]
For now, and until we find a better solution, yes.

**Comment by jonludlam** [comment]
For sending this command line over what's essentially a noisy channel, they could be encoded. For example, we're currently sending key=value pairs over the xen cmdline in xenstore, but we could choose to prefix those that we want the unikernel to be aware of - e.g. mirage:key=value, and we nuke anything that doesn't look like that. That'd be changes only to mirage-bootvar-xen and to whatever tool was being used to set the args.

**Comment by jonludlam** [comment]
OTOH, it'd be nice to share the same arg parsing stuff as the unix backend uses, so have something more like:
```
root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH 3 mirage_args="x y z"
```
and then splitting that as argv is split and passing that directly to cmdliner seems reasonable too.

**Comment by Drup** [comment]
And how do you access the non-mirage parameters ?

**Comment by jonludlam** [comment]
Do you want to?

**Comment by Drup** [comment]
I'm pretty sure people will want to at some point, yes.

**Comment by jonludlam** [comment]
After all, the purpose here is to ignore the parameters you _dont_ have any say over

**Comment by jonludlam** [comment]
Thinking again about this, the case I had in mind for ignoring some of the parameters was that XenServer exposes an API to _append_ arguments to a list extracted from grub. For launching unikernels we'll almost certainly have either full control over the arguments (by making a grub.conf, or by 'xe vm-param-set PV-args=...' on XS or by xl.conf), or no control at all (e.g. qubes). I would have thought the common case would be full control.

However, I guess the existence of _any_ Xen system under which you can't control the command line _requires_ us to be able to ignore it, otherwise we simply can't boot. The question is, do you compile your unikernel in advance to ignore it, or use some runtime magic to figure it out?

**Comment by talex5** [comment]
Actually, Qubes does allow you to append extra arguments. But you can't remove the default ones.

**Comment by jonludlam** [comment]
How does this sound: The command line is split up by whitespace (taking care to treat quotations correctly), and if one of those arguments starts with "mirage_args=" then we treat that as the argument string. If not, we keep the current behaviour. So for example:

```
root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH 3 mirage_args="x=y z=z"
```
comes out of bootvar looking like `['x','y'; 'z','z']`, but
```
root=/dev/mapper/dmroot console=hvc0
```
comes out looking as like `['root','/dev/mapper/dmroot'; 'console','hvc0']`. To fix Qubes, it's just a case of appending the argument `mirage_args=`. Would that work?

**Comment by talex5** [comment]
It's better than what we have now. But it still means the firewall will crash by default until the user changes the settings, which isn't ideal.

**Comment by jonludlam** [comment]
@hannesm @talex5 OK, how about the original idea of prefixing each key with something then? `mirage:foo=bar` or the like?

**Comment by jonludlam** [comment]
Actually, thinking some more, I think I still prefer the magic string option (though we could make it a bit more obscure if you think it's too easy to accidentally invoke it). It means the 'normal' case is less horrible, and the unusual case of Qubes is the only one where the magic has to be applied. @talex5 it it very bad if the default case doesn't work? Can this not be just documented as part of the installation instructions for Qubes? I would hope that the Qubes people see this as a defect and would fix it, in which case the magic case can disappear entirely, and we're left with a straightforward solution.

**Comment by talex5** [comment]
One of the benefits we advertise for unikernels is that they only include what you need and you can swap out any component with a replacement. So having a command-line parser that you don't need that always runs at startup, can't be overridden, and causes your unikernel to crash seems a little embarrassing.

So I think the first step here is to put the developer back in control, by letting them decide where the arguments come from. Then they can pass an empty array, call `Bootvar` and filter the output, or whatever makes sense for their case.

I think we should do this even if the Qubes and AWS problems get fixed.

(for now, I've made a fork of `Bootvar` that returns an empty arg list: https://github.com/talex5/mirage-bootvar-xen/tree/qubes and updated my README to tell users to pin that)

**Comment by samoht** [comment]
Fixed by #534. Now use:

```ocaml
register ~argv:no_argv [...]
```
